### DISCUSSION TITLE
Consider some new meanings for scaffold

### DISCUSSION THREAD

**Discussion by julie777** [pyscaffold/pyscaffold#592] [Dev Discussions]
I haven't seen this in any other discussions, so I thought I would bring it up to stimulate some thoughtful discussion.
Here are a couple quotes from the documentation.

### Now

> Do one thing and do it well
> It might sound old-fashioned, but we like this piece of good old UNIX wisdom. Pyscaffold creates a perfect project structure that compiles the best practices for Python packaging and has tons of useful defaults… that is it!

> PyScaffold is a project generator for bootstrapping high-quality Python packages

This makes me think of pyscaffold as a tool to scaffold/create a project ready to go. The user just has to add the code for their app/package. I think this is a great and necessary tool with only this one thing as the goal.

### Idea

In addition, when I was reading the discussion on pypa about adding pyscaffold I started thinking about another variation of scaffold. Because the python tool-set, especially for packaging has always been a moving target with new tools being created, new ideas being proposed, new PEPs being written, I can imaging Pyscaffold as the framework/scaffold that glues the ecosystem together.

In this role Pyscaffold would make it easy to swap out various tools in the automated tool chain that is the created project. This is much like the Unix idea of do one thing and do it well.  Think of how well shell pipelines work. Here are some examples:

- I write a new dependency manager and substitute it for pip in pyscaffold
- I swap out flit for some other way to package and upload to pypi
- I want to try behave instead of pytest for writing tests.
- I want to compare the performance of virtualenv with pyvenv or venv

I'm sure that you can see that I clumped two ideas together.
1. develop and test new tools easily in a full environment
2. allow users to choose tools that work best in their organization (Jenkins vs github actions)

The first thing that would come out of this a complete understanding and documenting of the process and tool chain. I am very impressed with the current documentation.  

This is probably more than enough for now.

**Comment by julie777** [comment]
Keeping putup pure and not trying to manage everything like Hatch. (Poetry was mentioned but I don't believe it does everything.)

I have spent many hours thinking about this, because you two seem to feel very strongly that putup should one create the project.
However, putup seems to be getting more and more dependent on other tools, even if they are extensions, they still fall under the umbrella of pyscaffold. As I mentioned above, a scaffold can be used for many things. In the physical world they use them to work on the outside of buildings. I once used scaffolding to paint the ceiling of my garage, quite a different kind of task.

Let's assume the putup just creates a project. What if pyscaffold had another tool that provided an easy to use command framework to plug in version control (I mean updating the product version for a release), and some amount of venv management, like the venv extension. This would be a way to handle all the simple cases without fully learning all the underlying tools. And it would provide a consistent interface for PyPA to experiment with various tools to determine what was best of breed and what worked well together.

A different aspect of tools and there usage, is what a tool provides when integrated into the project. Instead of putup having all the knowledge about all the tools and all the config for those tools, might it be better if a tool could plugin to pyscaffold and it would provide the best config file, and other information so we could make the best use of it. (In some ways I am talking about a variation of component based software.)

Something else for you to ponder. (when brainstorming all ideas should be captured. Later we can throw out the bad ones. I don't remember where that came from, but it fits what I have been doing.)

**Reply by FlorianWilhelm** [reply]
See [this reply](https://github.com/pyscaffold/pyscaffold/discussions/608#discussioncomment-2213453) first for my general thoughts. Then, let's further discuss this on Sunday?
